api/v1alpha4/clusterclass_webhook.go

```diff
@@ -132,6 +161,107 @@ func (in *ClusterClass) validate(old *ClusterClass) error {
 			)
 		}
 	}
+	return allErrs
+}
+
+func (in ClusterClass) validateRefTemplates() field.ErrorList {
+	var allErrs field.ErrorList
+
+	if !strings.HasSuffix(strings.ToLower(in.Spec.Infrastructure.Ref.Kind), "template") {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec", "infrastructure", "ref", "kind"),
+				in.Spec.Infrastructure.Ref.Kind,
+				"must be a template",
+			),
+		)
+	}
+	if !strings.HasSuffix(strings.ToLower(in.Spec.ControlPlane.Ref.Kind), "template") {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec", "controlplane", "ref", "kind"),
+				in.Spec.ControlPlane.Ref.Kind,
+				"must be a template",
+			),
+		)
+	}
+	for _, class := range in.Spec.Workers.MachineDeployments {
+		if !strings.HasSuffix(strings.ToLower(class.Template.Bootstrap.Ref.Kind), "template") {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "workers", "machineDeployments", "template", "bootstrap", "ref", "kind"),
+					class.Template.Bootstrap.Ref.Kind,
+					"must be a template",
+				),
+			)
+		}
+		if !strings.HasSuffix(strings.ToLower(class.Template.Infrastructure.Ref.Kind), "template") {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "workers", "machineDeployments", "template", "infrastructure", "ref", "kind"),
+					class.Template.Infrastructure.Ref.Kind,
+					"must be a template",
+				),
+			)
+		}
+	}
+	return allErrs
+}
+
+func (in ClusterClass) validateRefAPIVersions() field.ErrorList {
+	var allErrs field.ErrorList
+
+	if in.Spec.Infrastructure.Ref.APIVersion != "" {
+		if _, err := schema.ParseGroupVersion(in.Spec.Infrastructure.Ref.APIVersion); err != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "infrastructure", "ref", "apiVersion"),
+					in.Spec.Infrastructure.Ref.APIVersion,
+					fmt.Sprintf("must be a valid apiVerison. found error: %v", err),
```

Comment:
```suggestion
					fmt.Sprintf("must be a valid apiVerison: %v", err),
```
nit (I think we can build the error text as usual)